include/grpc/grpc_audit_logging.h

```diff
@@ -0,0 +1,71 @@
+//
+//
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#ifndef GRPC_GRPC_AUDIT_LOGGING_H
+#define GRPC_GRPC_AUDIT_LOGGING_H
+
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+
+namespace grpc_core {
+namespace experimental {
+
+// The base struct for audit context.
+typedef struct CoreAuditContext {
+ public:
+  absl::string_view rpc_method() const;
+  absl::string_view principal() const;
+  absl::string_view policy_name() const;
+  absl::string_view matched_rule() const;
+  bool authorized() const;
+} AuditContext;
+
+// This base class for audit logger implementations.
+class CoreAuditLogger {
+ public:
+  virtual void CoreLog(const CoreAuditContext& audit_context) = 0;
```

Comment:
This should just be Log(...), and we'll have an instance that has a `unique_ptr<grpc::AuditLogger> log_` that we call `log_->Log(...)` on.

Reply:
So it means this class defined in C-Core can have a dependence on the C++ class, right? I initially was about to do this but wasn't sure if it's fine to have such a reversed dependency.

Reply:
Oh. is the following flow correct?

```C++
// C-Core
class AuditLogger {
 public:
  virtual void Log() = 0;
};
```

```
// C++
class AuditLogger {
 public:
  virtual void Log() = 0;
};

// non-public API
class CoreLogger : public grpc_core::AuditLogger {
 public:
  void Log() { logger_->Log(); }
 private:
  std::unique_ptr<AuditLogger> logger_;
}
```